pkg/csi/csi_ops.go

```diff
@@ -0,0 +1,392 @@
+}
+
+func (c *cleanse) DeleteSnapshot(ctx context.Context, snapshotName string, namespace string, SnapshotGroupVersion *metav1.GroupVersionForDiscovery) error {
+	if c.dynCli == nil {
+		return fmt.Errorf("dynCli not initialized")
+	}
+	if SnapshotGroupVersion == nil || SnapshotGroupVersion.Version == "" {
+		return fmt.Errorf("snapshot group version not provided")
+	}
+	VolSnapGVR := schema.GroupVersionResource{Group: common.SnapGroupName, Version: SnapshotGroupVersion.Version, Resource: common.VolumeSnapshotResourcePlural}
+	return c.dynCli.Resource(VolSnapGVR).Namespace(namespace).Delete(ctx, snapshotName, metav1.DeleteOptions{})
+}
+
+//go:generate go run github.com/golang/mock/mockgen -destination=mocks/mock_api_version_fetcher.go -package=mocks . ApiVersionFetcher
+type ApiVersionFetcher interface {
+	GetCSISnapshotGroupVersion() (*metav1.GroupVersionForDiscovery, error)
+}
+
+type apiVersionFetch struct {
+	kubeCli kubernetes.Interface
+}
+
+func (p *apiVersionFetch) GetCSISnapshotGroupVersion() (*metav1.GroupVersionForDiscovery, error) {
+	if p.kubeCli == nil {
+		return nil, fmt.Errorf("kubeCli not initialized")
+	}
+	groups, _, err := p.kubeCli.Discovery().ServerGroupsAndResources()
+	if err != nil {
+		return nil, err
+	}
+	for _, group := range groups {
+		if group.Name == common.SnapGroupName {
+			return &group.PreferredVersion, nil
+		}
+	}
+	return nil, fmt.Errorf("Snapshot API group not found")
+}
+
+//go:generate go run github.com/golang/mock/mockgen -destination=mocks/mock_data_validator.go -package=mocks . DataValidator
+type DataValidator interface {
+	FetchPodData(podName string, podNamespace string) (string, error)
+}
+
+type validateData struct {
+	kubeCli kubernetes.Interface
+}
+
+func (p *validateData) FetchPodData(podName string, podNamespace string) (string, error) {
+	if p.kubeCli == nil {
+		return "", fmt.Errorf("kubeCli not initialized")
+	}
+	stdout, _, err := kankube.Exec(p.kubeCli, podNamespace, podName, "", []string{"sh", "-c", "cat /data/out.txt"}, nil)
+	return stdout, err
+}
+
+//go:generate go run github.com/golang/mock/mockgen -destination=mocks/mock_port_forwarder.go -package=mocks . PortForwarder
+type PortForwarder interface {
+	FetchRestConfig() (*rest.Config, error)
+	PortForwardAPod(req *types.PortForwardAPodRequest) error
+}
+
+type portforward struct{}
+
+func (p *portforward) PortForwardAPod(req *types.PortForwardAPodRequest) error {
+	path := fmt.Sprintf("/api/v1/namespaces/%s/pods/%s/portforward",
+		req.Pod.Namespace, req.Pod.Name)
+	hostIP := strings.TrimLeft(req.RestConfig.Host, "https:/")
+
+	transport, upgrader, err := spdy.RoundTripperFor(req.RestConfig)
+	if err != nil {
+		return err
+	}
+
+	dialer := spdy.NewDialer(upgrader, &http.Client{Transport: transport}, http.MethodPost, &url.URL{Scheme: "https", Path: path, Host: hostIP})
+	fw, err := pf.New(dialer, []string{fmt.Sprintf("%d:%d", req.LocalPort, req.PodPort)}, req.StopCh, req.ReadyCh, &req.OutStream, &req.ErrOutStream)
```

Comment:
Probably helpful to validate early on that the local port is available . Otherwise you might go through the whole process of creating a clone of the pvc and creating the file browser app , only to find that the local port is already used.  We could file a separate github issue as enhancement to track this .